### PR TITLE
Fix UBSan null-pointer UB in vec256 partial-load Vec::loadu (#181510)

### PR DESCRIPTION
**Pull request by SongTonyLi**
## What

`Vectorized<T>::loadu(ptr, count)` in the vec256 specializations calls `std::memcpy(tmp_values, ptr, count * sizeof(T))` unconditionally in its partial-load branch. When `ptr == nullptr` and `count == 0` (which happens when a zero-numel tensor's `data_ptr()` is passed in), this is UB:

```
aten/src/ATen/cpu/vec/vec256/vec256_float.h:125:21: runtime error:
null pointer passed as argument 2, which is declared to never be null
```

The repro from #181510:

```python
m = torch.nn.utils.weight_norm(torch.nn.Linear(0, 1))
m(torch.empty((1, 0), dtype=torch.float32))
```

walks this chain:

```
weight_norm_first_dim_kernel  (v_data=nullptr, N=0)
  -> vec::map_reduce_all       (size < Vec::size() branch in
                                 functional_base.h:250-251)
  -> Vectorized<float>::loadu(nullptr, 0)
  -> memcpy(tmp_values, nullptr, 0)   // UB
```

## Fix

Guard the `memcpy` with `if (count > 0)`:

```diff
-    std::memcpy(
-        tmp_values, reinterpret_cast<const float*>(ptr), count * sizeof(float));
+    if (count > 0) {
+      std::memcpy(
+          tmp_values, reinterpret_cast<const float*>(ptr), count * sizeof(float));
+    }
```

`tmp_values` is already zero-initialized by the loop above (intentional, not `={0}`, see #32502 for why). When `count == 0` there are no source bytes to copy, and the function returns the same all-zero vector the partial-load branch was already producing on every successful call. No behavior change for valid inputs.

The same guard goes on all 9 `loadu(ptr, count)` overloads in `aten/src/ATen/cpu/vec/vec256/`:

| File | Class | Guard |
|------|-------|-------|
| `vec256_float.h` | `Vectorized<float>` | `if (count > 0)` |
| `vec256_double.h` | `Vectorized<double>` | `if (count > 0)` |
| `vec256_16bit_float.h` | `Vectorized16<T>` (BFloat16, Half) | `if (count > 0 && count < size())` |
| `vec256_complex_float.h` | `Vectorized<c10::complex<float>>` | `if (count > 0)` |
| `vec256_complex_double.h` | `Vectorized<c10::complex<double>>` | `if (count > 0)` |
| `vec256_int.h` | `Vectorized<int64_t>` | `if (count > 0)` |
| `vec256_int.h` | `Vectorized<int32_t>` | `if (count > 0)` |
| `vec256_int.h` | `Vectorized<int16_t>` | `if (count > 0)` |
| `vec256_int.h` | `Vectorized8<T>` | `if (count > 0)` |

The `vec256_16bit_float.h` overload uses the stricter `count > 0 && count < size()` guard to satisfy gcc's `-Werror=array-bounds` analyzer. `count` is declared `int16_t` there, and gcc proved that with only the `count > 0` guard, an out-of-contract caller passing `count > size()` (e.g. `count == 17` with `size() == 16`) could form a 33-byte memcpy into the 32-byte `tmp_values` buffer. Adding `count < size()` lets the analyzer prove the memcpy stays within bounds. For in-contract callers (`count` in `[0, size()]`) the behavior is unchanged: `count == size()` short-circuits in the early-return above; `count` in `[0, size())` enters the partial-load branch and now harmlessly skips the memcpy when 0 (returning the zero-initialized `tmp_values`); `count > size()` is the previously-unsafe case where we now skip the memcpy instead of buffer-overflowing.

The vec512 paths use masked-load intrinsics (`_mm512_maskz_loadu_*`) and don't have this pattern.

## Relationship to #181665

#181665 already proposes a fix at the kernel level: early-exit in `weight_norm_kernel` and `weight_norm_backward_kernel` when `v.numel() == 0`, with `norm` / `grad_g` filled with zeros via `std::fill_n`. Its description calls this "the minimum fix" per @albanD's response on the issue. This PR is not a critique of that approach, just an alternative at a different layer.

Two reasons I picked the SIMD primitive layer:

1. `vec::map_reduce_all` has more than one production caller. A `git grep map_reduce_all aten/` finds two CPU kernels using it: `WeightNormKernel.cpp:32` (the one in #181510) and `LogSoftmaxKernelImpl.h:63`. Both pass `data + i * size` as the pointer; both would hit the same `loadu(nullptr, 0)` UB if that size is 0 and the underlying tensor's `data_ptr()` is null. The kernel-level fix covers `WeightNorm`. Patching `loadu` covers both, and any future caller that goes through the same `size < Vec::size()` branch.
2. `loadu` has no semantic decision to make, only a `memcpy` precondition to honor. Guarding `map_reduce_all` itself wouldn't work because `ReduceOp` is generic and has no universal identity; a forced zero-size return would silently wrong-answer non-additive reductions like max/min/product. The `count == 0` case in `loadu` is unambiguous: load zero elements, and the function already zero-initializes its return buffer.

The two fixes aren't mutually exclusive. #181665 documents the empty-tensor semantic for `weight_norm`; this PR removes the underlying UB at the primitive. Up to maintainers which (or both) to land.

## Tests

Two tests added in separate commits:

1. **C++** in `aten/src/ATen/test/vec_test_all_types.cpp` (`MemoryAllPatched.LoaduPartialNullPointer`). Calls `vec::loadu(nullptr, 0)` directly and asserts the returned vector is all zero. Runs across a custom `LoaduPatchedTypes` list (`vfloat, vdouble, vcomplex, vcomplexDbl, vBFloat16, vHalf, vlong, vint, vshort`) covering 8 of the 9 patched specializations. Plain int8/uint8 (`Vectorized8<T>` in `vec256_int.h`) doesn't have a typed-test alias in `vec_test_all_types.h`; the Python test below covers those via `sum()` over zero-numel int8/uint8 tensors. The `vqint8/vquint8/vqint` types are intentionally excluded — they route through `vec256_qint.h`, which has its own loadu overloads not touched by this PR.
2. **Python** in `test/test_nn.py` (`test_loadu_partial_null_pointer_zero_numel`). Mirrors #181665's `test_weight_norm_empty_input` style. Float family uses the original `weight_norm(nn.Linear(0, 1))` repro; complex and int dtypes use `sum()` on a zero-numel tensor.

Both tests are UBSan-only: `memcpy(_, nullptr, 0)` is UB by the standard but produces no observable runtime behavior on glibc, so a regression is only caught under a UBSan-instrumented build.

## Reproduction

UBSan harness (`-fsanitize=undefined -fno-sanitize-recover=all`) calling `Vectorized<T>::loadu(nullptr, 0)` directly across all 11 dtypes / 9 specializations:

| dtype | Specialization | Pre-patch UBSan | Post-patch |
|-------|----------------|-----------------|------------|
| `float` | `Vectorized<float>` | `vec256_float.h:125:21` | OK, all-zero vector |
| `double` | `Vectorized<double>` | `vec256_double.h:94:9` | OK, all-zero vector |
| `bfloat16` | `Vectorized16<BFloat16>` | `vec256_16bit_float.h:261:29` | OK, all-zero vector |
| `float16` | `Vectorized16<Half>` | `vec256_16bit_float.h:261:29` | OK, all-zero vector |
| `complex64` | `Vectorized<c10::complex<float>>` | `vec256_complex_float.h:175:9` | OK, all-zero vector |
| `complex128` | `Vectorized<c10::complex<double>>` | `vec256_complex_double.h:110:9` | OK, all-zero vector |
| `int64` | `Vectorized<int64_t>` | `vec256_int.h:120:29` | OK, all-zero vector |
| `int32` | `Vectorized<int32_t>` | `vec256_int.h:270:29` | OK, all-zero vector |
| `int16` | `Vectorized<int16_t>` | `vec256_int.h:570:29` | OK, all-zero vector |
| `int8` | `Vectorized8<int8_t>` | `vec256_int.h:918:29` | OK, all-zero vector |
| `uint8` | `Vectorized8<uint8_t>` | `vec256_int.h:918:29` | OK, all-zero vector |

Additional check on the post-patch float harness:
- `valgrind --tool=memcheck`: `ERROR SUMMARY: 0 errors from 0 contexts`, `All heap blocks were freed -- no leaks are possible`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01

### CODE REVIEW

**Comment by albanD** [comment]
As mentioned in the issue, this is way too involved and I don't think we'll want to investigate such potential performance impact. We should go with #181665 once it's simplified further.

**Comment by SongTonyLi** [comment]
@albanD Thank you for your response!

Indeed, I also thought about the complexity of the fix layer at the beginning until I audited what else calls these primitives. The kernel-level fix in #181665 covers WeightNormKernel. However, I digged a little deeper, and  vec::map2_reduce_all / map3_reduce_all / reduce_all have 25 call sites across 11 kernel files in aten/src/ATen/native/cpu/. Most are guarded upstream (numel==0 early-exit, parallel_reduce empty-range, etc.).

However, two are not:

  Kernel: group_norm_kernel.cpp:546, 552
  Call site: ComputeInternalGradients (backward)
  Why unguarded: group_norm.cpp:104 has no numel guard; parallel_for(0, N*C) runs with HxW=0

  Kernel: layer_norm_kernel.cpp:264, 270, 273, 280, 380, 386
  Call site: LayerNormBackwardKernelImpl (backward)
  Why unguarded: layer_norm.cpp:186 checks M > 0 but not N > 0

and they are trivially reachable from user code:

```
  # group_norm backward — triggers UBSan
  F.group_norm(torch.randn(1, 4, 0, requires_grad=True), 4).sum().backward()

  # layer_norm backward — triggers UBSan
  F.layer_norm(torch.randn(2, 0, requires_grad=True), [0]).sum().backward()
```

A regression (PoC) test covering both paths is added in this PR added in commit [15ac583](https://github.com/pytorch/pytorch/pull/181681/commits/15ac583fa1dac679f42ad019bc4d6245c30a302c) (test/test_nn.py:test_loadu_partial_null_pointer_backward):

```
  def test_loadu_partial_null_pointer_backward(self):
      # Backward-path regression test for #181510. The forward-only test
      # above covers weight_norm / sum; these backward kernels also reach
      # vec::loadu(nullptr, 0) through map2_reduce_all / map3_reduce_all
      # when an inner dimension is 0 and the outer dimension is > 0.

      # layer_norm backward: M=2, N=0 dispatches into
      # LayerNormBackwardKernelImpl with size=N=0 and nullptr data.
      x = torch.randn(2, 0, requires_grad=True)
      F.layer_norm(x, [0]).sum().backward()
      self.assertEqual(x.grad.shape, x.shape)

      # group_norm backward: N*C=4, HxW=0 dispatches parallel_for(0, 4)
      # with size=0 and nullptr inner data.
      x = torch.randn(1, 4, 0, requires_grad=True)
      F.group_norm(x, 4).sum().backward()
      self.assertEqual(x.grad.shape, x.shape)
```

UBSan verification (three-way, same container/toolchain used for the existing C++ harness):
```
  Build: Unpatched main
    group_norm bwd PoC: UBSan: null pointer passed as argument 2 at vec256_float.h:126 via ComputeInternalGradients → map2_reduce_all → loadu
    weight_norm fwd PoC: UBSan: same at vec256_float.h:126 via map_reduce_all

  Build: This PR (vec256 guards)
    group_norm bwd PoC: Clean pass
    weight_norm fwd PoC: Clean pass

  Build: #181665 (WeightNormKernel only)
    group_norm bwd PoC: UBSan: identical trace — group_norm_kernel.cpp:546 still hits loadu(nullptr, 0)
    weight_norm fwd PoC: Clean pass (guarded at kernel level)
```

  Reproduction environment

  The three-way UBSan verification was performed in a Docker container (secb-tools:pytorch.issue-162476-patch, x86_64) built on Ubuntu 20.04 with clang 18.1.8 and Python 3.11.4. PyTorch is built from source (commit b87ae5b, pre-patch baseline) in Release mode with ASan enabled (-fsanitize=address -fno-omit-frame-pointer -g). To add UBSan without a full rebuild, we recompiled only the 6 kernel object files that inline vec::loadu via the reduce/map_reduce paths layer_norm_kernel.cpp, group_norm_kernel.cpp, and WeightNormKernel.cpp (DEFAULT + AVX2 variants) with -fsanitize=address,undefined -fno-sanitize-recover=undefined, then relinked libtorch_cpu.so. To reproduce:

```
  docker pull songtli/pytorch_fix_181510:latest
  docker run --rm -it songtli/pytorch_fix_181510:latest bash

  # Inside the container:
  LD_PRELOAD=/usr/local/lib/clang/18/lib/x86_64-unknown-linux-gnu/libclang_rt.asan.so \
  ASAN_OPTIONS=detect_leaks=0 \
  UBSAN_OPTIONS="halt_on_error=1:print_stacktrace=1:silence_unsigned_overflow=1" \
  python -c "
  import torch, torch.nn.functional as F
  F.group_norm(torch.randn(1, 4, 0, requires_grad=True), 4).sum().backward()
  print('OK')
  "
```
  On unpatched main this halts with runtime error: null pointer passed as argument 2 at vec256_float.h:126. With this PR's vec256 guards applied it prints OK. The container image is available at docker.io/songtli/pytorch_fix_181510 for independent verification.
